File: react-frontend/backend/AdminService.java
package com.mealbasket.service;

import com.mealbasket.model.User;
import com.mealbasket.model.Product;
import com.mealbasket.repository.UserRepository;
import com.mealbasket.repository.ProductRepository;
import com.mealbasket.dto.DashboardStatsDTO;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.stereotype.Service;

import java.util.List;
import java.util.Map;
import java.util.HashMap;
import java.util.stream.Collectors;

@Service
public class AdminService {

    @Autowired
    private UserRepository userRepository;

    @Autowired
    private ProductRepository productRepository;

    /**
     * Get all users from the database
     */
    public List<User> getAllUsers() {
        System.out.println("AdminService: Fetching all users from database");
        List<User> users = userRepository.findAll();
        
        // Add vendor-specific information for vendors
        users = users.stream().map(user -> {
            if ("VENDOR".equals(user.getRole())) {
                // Get product count for this vendor
                List<Product> vendorProducts = productRepository.findByVendorId(user.getId());
                user.setProductCount(vendorProducts.size());
                user.setIsConnected(vendorProducts.size() > 0);
                user.setLastActivity(vendorProducts.size() > 0 ? "Active" : "No products yet");
            } else if ("USER".equals(user.getRole())) {
                user.setIsConnected(true);
                user.setLastActivity("Active");
            } else if ("ADMIN".equals(user.getRole())) {
                user.setIsConnected(true);
                user.setLastActivity("Managing System");
            }
            return user;
        }).collect(Collectors.toList());
        
        System.out.println("AdminService: Found " + users.size() + " users");
        return users;
    }

    /**
     * Get all vendors specifically
     */
    public List<User> getAllVendors() {
        System.out.println("AdminService: Fetching all vendors from database");
        List<User> vendors = userRepository.findByRole("VENDOR");
        
        // Add vendor-specific information
        vendors = vendors.stream().map(vendor -> {
            List<Product> vendorProducts = productRepository.findByVendorId(vendor.getId());
            vendor.setProductCount(vendorProducts.size());
            vendor.setIsConnected(vendorProducts.size() > 0);
            vendor.setLastActivity(vendorProducts.size() > 0 ? "Active" : "No products yet");
            return vendor;
        }).collect(Collectors.toList());
        
        System.out.println("AdminService: Found " + vendors.size() + " vendors");
        return vendors;
    }

    /**
     * Get dashboard statistics
     */
    public DashboardStatsDTO getDashboardStats() {
        System.out.println("AdminService: Calculating dashboard statistics");
        
        List<User> allUsers = userRepository.findAll();
        List<User> vendors = userRepository.findByRole("VENDOR");
        List<User> regularUsers = userRepository.findByRole("USER");
        List<User> admins = userRepository.findByRole("ADMIN");
        List<Product> allProducts = productRepository.findAll();
        
        // Count connected users and active vendors
        long connectedUsers = allUsers.stream()
            .filter(User::getIsConnected)
            .count();
        
        long activeVendors = vendors.stream()
            .filter(User::getIsConnected)
            .count();
        
        // Calculate total revenue (sum of all product prices)
        double totalRevenue = allProducts.stream()
            .mapToDouble(product -> product.getPrice() != null ? product.getPrice() : 0.0)
            .sum();
        
        DashboardStatsDTO stats = new DashboardStatsDTO();
        stats.setTotalUsers(allUsers.size());
        stats.setTotalVendors(vendors.size());
        stats.setTotalRegularUsers(regularUsers.size());
        stats.setTotalAdmins(admins.size());
        stats.setTotalProducts(allProducts.size());
        stats.setTotalOrders(0); // Will be implemented when orders are added
        stats.setTotalRevenue(totalRevenue);
        stats.setConnectedUsers((int) connectedUsers);
        stats.setActiveVendors((int) activeVendors);
        
        System.out.println("AdminService: Statistics calculated - Users: " + stats.getTotalUsers() + 
                          ", Vendors: " + stats.getTotalVendors() + 
                          ", Products: " + stats.getTotalProducts());
        
        return stats;
    }

    /**
     * Get all products from all vendors
     */
    public List<Product> getAllProducts() {
        System.out.println("AdminService: Fetching all products from database");
        List<Product> products = productRepository.findAll();
        
        // Add vendor information to each product
        products = products.stream().map(product -> {
            if (product.getVendorId() != null) {
                User vendor = userRepository.findById(product.getVendorId()).orElse(null);
                if (vendor != null) {
                    product.setVendorName(vendor.getBusinessName() != null ? 
                                       vendor.getBusinessName() : vendor.getName());
                }
            }
            return product;
        }).collect(Collectors.toList());
        
        System.out.println("AdminService: Found " + products.size() + " products");
        return products;
    }

    /**
     * Delete a user by ID
     */
    public boolean deleteUser(Long userId) {
        System.out.println("AdminService: Deleting user with ID: " + userId);
        
        try {
            if (userRepository.existsById(userId)) {
                // First, delete all products associated with this user if they're a vendor
                User user = userRepository.findById(userId).orElse(null);
                if (user != null && "VENDOR".equals(user.getRole())) {
                    List<Product> vendorProducts = productRepository.findByVendorId(userId);
                    productRepository.deleteAll(vendorProducts);
                    System.out.println("AdminService: Deleted " + vendorProducts.size() + " products for vendor " + userId);
                }
                
                // Delete the user
                userRepository.deleteById(userId);
                System.out.println("AdminService: Successfully deleted user " + userId);
                return true;
            } else {
                System.out.println("AdminService: User " + userId + " not found");
                return false;
            }
        } catch (Exception e) {
            System.err.println("AdminService: Error deleting user " + userId + ": " + e.getMessage());
            e.printStackTrace();
            return false;
        }
    }

    /**
     * Update user role
     */
    public User updateUserRole(Long userId, String newRole) {
        System.out.println("AdminService: Updating role for user " + userId + " to " + newRole);
        
        try {
            User user = userRepository.findById(userId).orElse(null);
            if (user != null) {
                // Validate role
                if (!"USER".equals(newRole) && !"VENDOR".equals(newRole) && !"ADMIN".equals(newRole)) {
                    System.out.println("AdminService: Invalid role: " + newRole);
                    return null;
                }
                
                String oldRole = user.getRole();
                user.setRole(newRole);
                
                // Reset connection status based on new role
                if ("VENDOR".equals(newRole)) {
                    List<Product> vendorProducts = productRepository.findByVendorId(userId);
                    user.setProductCount(vendorProducts.size());
                    user.setIsConnected(vendorProducts.size() > 0);
                    user.setLastActivity(vendorProducts.size() > 0 ? "Active" : "No products yet");
                } else if ("USER".equals(newRole)) {
                    user.setIsConnected(true);
                    user.setLastActivity("Active");
                } else if ("ADMIN".equals(newRole)) {
                    user.setIsConnected(true);
                    user.setLastActivity("Managing System");
                }
                
                userRepository.save(user);
                System.out.println("AdminService: Successfully updated role for user " + userId + 
                                  " from " + oldRole + " to " + newRole);
                return user;
            } else {
                System.out.println("AdminService: User " + userId + " not found");
                return null;
            }
        } catch (Exception e) {
            System.err.println("AdminService: Error updating user role: " + e.getMessage());
            e.printStackTrace();
            return null;
        }
    }
}

File: react-frontend/src/services/adminService.js
// Admin Service for backend API integration
class AdminService {
  constructor() {
    this.baseURL = 'http://localhost:8081/api/admin';
    this.productsURL = 'http://localhost:8081/api/products';
  }

  // Get authentication headers
  getAuthHeaders() {
    const token = localStorage.getItem('token');
    return {
      'Content-Type': 'application/json',
      'Authorization': `Bearer ${token}`
    };
  }

  // Get all users and vendors from backend
  async getAllUsers() {
    try {
      console.log('AdminService: Fetching all users from backend...');
      const response = await fetch(`${this.baseURL}/users`, {
        method: 'GET',
        headers: this.getAuthHeaders()
      });

      if (!response.ok) {
        let bodyText = '';
        try {
          bodyText = await response.text();
        } catch (e) {
          bodyText = '';
        }
        const msg = `HTTP ${response.status} ${response.statusText}${bodyText ? ` - ${bodyText}` : ''}`;
        const err = new Error(msg);
        err.status = response.status;
        throw err;
      }

      const data = await response.json();
      console.log('AdminService: Successfully fetched users:', data);
      return data;
    } catch (error) {
      console.error('AdminService: Error fetching users:', error);
      throw error;
    }
  }

  // Get all vendors specifically
  async getAllVendors() {
    try {
      console.log('AdminService: Fetching all vendors from backend...');
      const response = await fetch(`${this.baseURL}/vendors`, {
        method: 'GET',
        headers: this.getAuthHeaders()
      });

      if (!response.ok) {
        let bodyText = '';
        try {
          bodyText = await response.text();
        } catch (e) {
          bodyText = '';
        }
        const msg = `HTTP ${response.status} ${response.statusText}${bodyText ? ` - ${bodyText}` : ''}`;
        const err = new Error(msg);
        err.status = response.status;
        throw err;
      }

      const data = await response.json();
      console.log('AdminService: Successfully fetched vendors:', data);
      return data;
    } catch (error) {
      console.error('AdminService: Error fetching vendors:', error);
      throw error;
    }
  }

  // Get dashboard statistics
  async getDashboardStats() {
    try {
      console.log('AdminService: Fetching dashboard stats from backend...');
      const response = await fetch(`${this.baseURL}/stats`, {
        method: 'GET',
        headers: this.getAuthHeaders()
      });

      if (!response.ok) {
        throw new Error(`HTTP error! status: ${response.status}`);
      }

      const data = await response.json();
      console.log('AdminService: Successfully fetched stats:', data);
      return data;
    } catch (error) {
      console.error('AdminService: Error fetching stats:', error);
      throw error;
    }
  }

  // Get all products from all vendors
  async getAllProducts() {
    try {
      console.log('AdminService: Fetching all products from backend...');
      const response = await fetch(`${this.productsURL}`, {
        method: 'GET',
        headers: {
          'Content-Type': 'application/json'
        }
      });

      if (!response.ok) {
        throw new Error(`HTTP error! status: ${response.status}`);
      }

      const data = await response.json();
      console.log('AdminService: Successfully fetched products:', data);
      return data;
    } catch (error) {
      console.error('AdminService: Error fetching products:', error);
      throw error;
    }
  }

  // Get all orders from backend
  async getAllOrders() {
    try {
      console.log('AdminService: Fetching all orders from backend...');
      const response = await fetch(`${this.baseURL}/orders`, {
        method: 'GET',
        headers: this.getAuthHeaders()
      });

      if (!response.ok) {
        let bodyText = '';
        try {
          bodyText = await response.text();
        } catch (e) {
          bodyText = '';
        }
        const msg = `HTTP ${response.status} ${response.statusText}${bodyText ? ` - ${bodyText}` : ''}`;
        const err = new Error(msg);
        err.status = response.status;
        throw err;
      }

      const data = await response.json();
      console.log('AdminService: Successfully fetched orders:', data);
      return data;
    } catch (error) {
      console.error('AdminService: Error fetching orders:', error);
      throw error;
    }
  }

  // Get all ratings from backend
  async getAllRatings() {
    try {
      console.log('AdminService: Fetching all ratings from backend...');
      const response = await fetch(`${this.baseURL}/ratings`, {
        method: 'GET',
        headers: this.getAuthHeaders()
      });

      if (!response.ok) {
        let bodyText = '';
        try {
          bodyText = await response.text();
        } catch (e) {
          bodyText = '';
        }
        const msg = `HTTP ${response.status} ${response.statusText}${bodyText ? ` - ${bodyText}` : ''}`;
        const err = new Error(msg);
        err.status = response.status;
        throw err;
      }

      const data = await response.json();
      console.log('AdminService: Successfully fetched ratings:', data);
      return data;
    } catch (error) {
      console.error('AdminService: Error fetching ratings:', error);
      throw error;
    }
  }

  // Delete a user
  async deleteUser(userId) {
    try {
      console.log('AdminService: Deleting user:', userId);
      const response = await fetch(`${this.baseURL}/users/${userId}`, {
        method: 'DELETE',
        headers: this.getAuthHeaders()
      });

      if (!response.ok) {
        throw new Error(`HTTP error! status: ${response.status}`);
      }

      const data = await response.json();
      console.log('AdminService: Successfully deleted user:', data);
      return data;
    } catch (error) {
      console.error('AdminService: Error deleting user:', error);
      throw error;
    }
  }

  // Update user role
  async updateUserRole(userId, newRole) {
    try {
      console.log('AdminService: Updating user role:', userId, newRole);
      const response = await fetch(`${this.baseURL}/users/${userId}/role`, {
        method: 'PUT',
        headers: this.getAuthHeaders(),
        body: JSON.stringify({ role: newRole })
      });

      if (!response.ok) {
        throw new Error(`HTTP error! status: ${response.status}`);
      }

      const data = await response.json();
      console.log('AdminService: Successfully updated user role:', data);
      return data;
    } catch (error) {
      console.error('AdminService: Error updating user role:', error);
      throw error;
    }
  }

  // Fallback to localStorage if backend is not available
  async getUsersFromFallback() {
    console.log('AdminService: Using localStorage fallback for users...');
    const allUsers = JSON.parse(localStorage.getItem('allUsers') || '[]');
    return allUsers;
  }
}

const adminService = new AdminService();
export default adminService;
